### DISCUSSION TITLE
How make android app have root permission to access something?

### DISCUSSION THREAD

**Discussion by ccinsz** [beeware/briefcase#1303] [Q&A]
Hi, the app need to access something, but it don't have permission, like "com.chaquo.python.PyException: CalledProcessError: Command 'setenforce 0' returned non-zero exit status 1."
How make android app have root permission to access something?

**Comment by mhsmith** [comment]
For obvious reasons, you can't disable SELinux from within an app. You'll have to do it manually in a root shell.

Or if you explain *why* you want to disable SELinux, we might be able to suggest another solution.

**Reply by ccinsz** [reply]
I want to call subprocess.check_output("chmod 777 /dev/input/event3", shell=True), so the app can operate the device.

**Reply by ccinsz** [reply]
the app raises "com.chaquo.python.PyException: CalledProcessError: Command 'chmod 777 /dev/input/event3' returned non-zero exit status 1."
It needs the permission.

**Reply by mhsmith** [reply]
What is this device, and what do you want to do with it? If there are standard Android APIs for what you're trying to accomplish, you'll find that much easier than direct device file access.

**Reply by ccinsz** [reply]
This device is the mobile phone. I want to access some parameters of the mobile phone system through this app. I'm making an app for the first time, so I don't how to do it the better.

**Reply by freakboy3742** [reply]
We understand you're on a mobile phone - we want to know why you're trying to access the `/dev/input/event3` device. Although Android does have Linux infrastructure internally, you generally don't interact with devices in that way on Android - and you generally don't use subprocess either. There's almost certainly a better way to do what you're trying to do - and even if you do need direct device access for some reason, knowing what the `/dev/input/event3` device *is* will help us guide you on what permissions might be missing.

**Reply by ccinsz** [reply]
If don't use subprocess, what can I use to interact with devices?

/dev/input/event3 is the phone. I want to use some commands, like "sendevent /dev/input/event3 3 57 2727", and  /dev/input/event3 needs the permission.

**Reply by mhsmith** [reply]
> I want to access some parameters of the mobile phone system through this app.

What parameters? What events? None of the commands you've provided return any obvious results on Google, and we can't help you unless you properly explain what you're trying to achieve.

**Reply by ccinsz** [reply]
I want to use the app to call up other app and do automated testing. For example, the app can call up "Settings" and click "Storage".

"sendevent /dev/input/event3 3 57 2727" is one command of click events.

**Reply by mhsmith** [reply]
If all you want to do is open a settings page, then you can do that using an `Intent`, as shown [here](https://stackoverflow.com/questions/19517417). The Toga equivalent of `startActivityForResult` is `intent_result`, which is described [here](https://github.com/beeware/toga/issues/1798), and the names of the different settings pages are all listed [here](https://developer.android.com/reference/android/provider/Settings).

However, this doesn't allow you programmatically *change* any of the settings. And if even if you could generate click events, the positions of the clicks wouldn't be portable between devices anyway. So once you get root access (which there's no standard way of doing), then you might as well change the setting directly at the command line.

Anyway, that's all I know about this area, and this doesn't have anything to do with BeeWare specifically, so if you want more details you'll have to ask in a more general Android programming forum. But I warn you, this is a pretty advanced thing to be attempting in your first mobile app.

**Reply by ccinsz** [reply]
Got it. Thank you for your help.

**Reply by ccinsz** [reply]
I find the Intent is not easy for python to get. How to make the right intent?

**Reply by mhsmith** [reply]
There's an example of creating an Intent [here](https://github.com/beeware/toga/pull/1158#issuecomment-1446196973). It's a different kind of Intent, but you should be able to adapt it.
